### PR TITLE
AC_WPNav:get_wp_destination_loc does not set altitude frame

### PR DESCRIPTION
**Pull request by arshPratap**
This PR solves the altitude frame issue raised here #14621.Thanks!

### CODE REVIEW

File: libraries/AC_WPNav/AC_WPNav.cpp [hunk 1]
```diff
@@ -197,6 +197,12 @@ bool AC_WPNav::get_wp_destination(Location& destination) const
     if (!AP::ahrs().get_origin(destination)) {
         return false;
     }
+    if (_terrain_alt==true) {
+        destination.change_alt_frame(Location::AltFrame::ABOVE_TERRAIN);
+    }
```
Comment (amilcarlucas) [inline libraries/AC_WPNav/AC_WPNav.cpp:202]: can you please use "} else {" instead

**Review by amilcarlucas — CHANGES_REQUESTED** [review]

File: libraries/AC_WPNav/AC_WPNav.cpp (hunk 1 above)
Comment (arshPratap) [reply libraries/AC_WPNav/AC_WPNav.cpp:202]: Definitely, Sir!!Thank You for pointing it out

**Comment by arshPratap** [comment]
@amilcarlucas Sir, I have made the requested changes. Please take a look. Thank You!!

**Comment by arshPratap** [comment]
@rmackay9  sir , it would be great if you take a look.Thanks!!

**Comment by peterbarker** [comment]
@arshPratap just to let you know, this one will get looked at after the s-curve navigation stuff goes in...

**Comment by hendjoshsr71** [comment]
This also effects the following modes auto, guided, & RTL through `get_oa_wp_destination()` which calls `get_wp_destination()`

Here are a few screenshots from SITL with terrain following active. 

Current change:
image


My Suggested change:
image

**Review by hendjoshsr71 — COMMENTED** [review]
I don't believe this is the correct change as you will get an additional altitude offset. Sorry for the order of the posts I'm learning how to use suggested changes.

File: libraries/AC_WPNav/AC_WPNav.cpp [hunk 2]
```diff
@@ -197,6 +197,11 @@ bool AC_WPNav::get_wp_destination(Location& destination) const
     if (!AP::ahrs().get_origin(destination)) {
         return false;
     }
+    if (_terrain_alt==true) {
+        destination.change_alt_frame(Location::AltFrame::ABOVE_TERRAIN);
+    } else {
+        destination.change_alt_frame(Location::AltFrame::ABOVE_ORIGIN);
+    }
     destination.offset(dest.x*0.01f, dest.y*0.01f);
     destination.alt += dest.z;
```
[the remark this answers is not in the document]
Comment (arshPratap) [reply libraries/AC_WPNav/AC_WPNav.cpp]: Thanks @hendjoshsr71 for looking into this PR and suggesting these changes.The suggested changes have been reflected in to the commit.Please do take a look !Thanks

**Comment by arshPratap** [comment]
The suggested changes have been reflected.
Thank you @peterbarker,@amilcarlucas and @hendjoshsr71 for taking out time to review this PR.

**Review by amilcarlucas — COMMENTED** [review]
How about this?

File: libraries/AC_WPNav/AC_WPNav.cpp [hunk 3]
```diff
@@ -199,6 +199,13 @@ bool AC_WPNav::get_wp_destination(Location& destination) const
     }
     destination.offset(dest.x*0.01f, dest.y*0.01f);
     destination.alt += dest.z;
+    
+    if (_terrain_alt == true) {
+        destination.set_alt_cm(destination.alt, Location::AltFrame::ABOVE_TERRAIN);
+    } else {
+        destination.set_alt_cm(destination.alt, Location::AltFrame::ABOVE_ORIGIN);
+    }
+
```
[the remark this answers is not in the document]
Comment (arshPratap) [reply libraries/AC_WPNav/AC_WPNav.cpp]: Ah thanks for this @amilcarlucas sir ! I have committed the changes.

**Comment by hendjoshsr71** [comment]
Since it will inevitably come up, next you should squash your commits to a single commit. https://ardupilot.org/dev/docs/git-interactive-rebase.html (This is the command line version, but I like my GUI and Sourcetree instead :) )

**Comment by arshPratap** [comment]
I have made the suggested changes. @hendjoshsr71 , @amilcarlucas please do take a look.If there are any more changes I would be happy to work on them too!
Thanks again

**Review by amilcarlucas — APPROVED** [review]
Git commit message and code-form look good to me.
Functionality wise .... no clue.

**Comment by hendjoshsr71** [comment]
@arshPratap 
So functionally this doesn't work and shows the importance of independent testing as I looked at this very quickly! Sorry, this was my mistake. This function is currently converting a vector referenced from EKF origin to altitude in above mean sea level or above terrain. 

This is my preferred solution with the code currently in master and could allow later factoring out of the return. But it is up to @rmackay9 as he has a fix for this in one of his branches currently. 

```cpp
bool AC_WPNav::get_wp_destination(Location& destination) const
{
    Vector3f dest = get_wp_destination();
    destination = Location(dest);
    if (_terrain_alt) {
        destination.set_alt_cm(dest.z, Location::AltFrame::ABOVE_TERRAIN);
    }
    return true;
}
```

However, using a PR I just submitted this could all become,
```cpp
bool AC_WPNav::get_wp_destination(Location& destination) const
{
    destination = Location(get_wp_destination(), _terrain_alt ? Location::AltFrame::ABOVE_TERRAIN : Location::AltFrame::ABOVE_ORIGIN);
    return true;
}
```

**Comment by hendjoshsr71** [comment]
Thanks for your effort here! I'm closing this as it got fixed with the scurves PR.

**Pull request closed without merging**